src/heimdall-tree/index.js

```diff
@@ -0,0 +1,243 @@
+import HeimdallNode from './node';
+import HeimdallLeaf from './leaf';
+import HashMap from '../shared/hash-map';
+import { normalizeTime } from '../shared/time';
+import { NULL_NUMBER } from '../shared/counter-store';
+import {
+  OP_START,
+  OP_STOP,
+  OP_RESUME,
+  OP_ANNOTATE
+} from '../shared/op-codes';
+
+/*
+Example Event Timeline and tree reconstruction
+
+As       Bs       Cs       Ce       Be       Ae
+|--------|--------|--------|--------|--------|
+   AB        BC       CC      CB        BA
+
+Tree
+A
+ \
+  B
+   \
+    C
+
+Leafy Tree
+A <- node
+ |_AB <- leaf
+ |_ B <- child node
+ | |_ BC
+ | |
+ | |_C
+ | | |_ CC
+ | |
+ | |_ CB
+ |
+ |_ BA
+
+*/
+
+function statsFromCounters(counterStore, counterCache) {
+  if (!counterStore || !counterCache) {
+    return null;
+  }
+
+  return counterStore.restoreFromCache(counterCache);
+}
+
+export default class HeimdallTree {
+  constructor(heimdall) {
```

Comment:
hmmm we should noodle through this.  I would expect us to be able to reuse this thing between in-process & parsing subsequent json files, but maybe not.